blockchain/pool.go

```diff
@@ -1,630 +1,383 @@
-		atomic.AddInt32(&pool.numPending, -1)
-		peer := pool.peers[peerID]
-		if peer != nil {
-			peer.decrPending(blockSize)
-		}
-	} else {
-		pool.Logger.Info("invalid peer", "peer", peerID, "blockHeight", block.Height)
-		pool.sendError(errors.New("invalid peer"), peerID)
-	}
-}
+	// Update the pool's maxPeerHeight if needed. Note that for updates it can only increase or left unchanged.
+	pool.updateMaxPeerHeight()
 
-// MaxPeerHeight returns the highest reported height.
-func (pool *BlockPool) MaxPeerHeight() int64 {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-	return pool.maxPeerHeight
+	return nil
 }
 
-// SetPeerHeight sets the peer's alleged blockchain height.
-func (pool *BlockPool) SetPeerHeight(peerID p2p.ID, height int64) {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	peer := pool.peers[peerID]
-	if peer != nil {
-		peer.height = height
-	} else {
-		peer = newBPPeer(pool, peerID, height)
-		peer.setLogger(pool.Logger.With("peer", peerID))
-		pool.peers[peerID] = peer
+// Stops the peer timer and deletes the peer. Recomputes the max peer height.
+func (pool *blockPool) deletePeer(peer *bpPeer) {
+	if peer == nil {
+		return
+	}
+	if peer.timeout != nil {
+		peer.timeout.Stop()
 	}
+	delete(pool.peers, peer.id)
 
-	if height > pool.maxPeerHeight {
-		pool.maxPeerHeight = height
+	if peer.height == pool.maxPeerHeight {
+		pool.updateMaxPeerHeight()
 	}
 }
 
-// RemovePeer removes the peer with peerID from the pool. If there's no peer
-// with peerID, function is a no-op.
-func (pool *BlockPool) RemovePeer(peerID p2p.ID) {
-	pool.mtx.Lock()
-	defer pool.mtx.Unlock()
-
-	pool.removePeer(peerID)
-}
+// Removes any blocks and requests associated with the peer and deletes the peer.
+// Also triggers new requests if blocks have been removed.
+func (pool *blockPool) removePeer(peerID p2p.ID, err error) {
+	peer := pool.peers[peerID]
+	if peer == nil {
+		return
+	}
+	pool.logger.Info("removing peer", "peerID", peerID, "error", err)
 
-func (pool *BlockPool) removePeer(peerID p2p.ID) {
-	for _, requester := range pool.requesters {
-		if requester.getPeerID() == peerID {
-			requester.redo(peerID)
+	// Reschedule the block requests made to the peer, or received and not processed yet.
+	// Note that some of the requests may be removed further down.
+	for h, block := range pool.peers[peerID].blocks {
+		if block == nil {
+			pool.numPending--
 		}
+		pool.rescheduleRequest(peerID, h)
 	}
 
-	peer, ok := pool.peers[peerID]
-	if ok {
-		if peer.timeout != nil {
-			peer.timeout.Stop()
+	oldMaxPeerHeight := pool.maxPeerHeight
+	// Delete the peer. This operation may result in the pool's maxPeerHeight being lowered.
+	pool.deletePeer(peer)
+
+	// Check if the pool's maxPeerHeight has been lowered.
+	// This may happen if the tallest peer has been removed.
+	if oldMaxPeerHeight > pool.maxPeerHeight {
+		// Remove any planned requests for heights over the new maxPeerHeight.
+		for h := range pool.plannedRequests {
+			if h > pool.maxPeerHeight {
```

Comment:
There is a lot of logic that is `maxPeerHeight` related. I am still not convinced that it's a good idea managing this variable.